nav2_waypoint_follower/src/waypoint_follower.cpp

```diff
@@ -304,4 +379,58 @@ WaypointFollower::goalResponseCallback(
   }
 }
 
+std::vector<geometry_msgs::msg::PoseStamped>
+WaypointFollower::convertGPSPoses2MapPoses(
+  const std::vector<nav2_msgs::msg::OrientedNavSatFix> & gps_poses,
+  const rclcpp_lifecycle::LifecycleNode::SharedPtr & parent_node,
+  const std::unique_ptr<nav2_util::ServiceClient<robot_localization::srv::FromLL>> & fromll_client)
+{
+  RCLCPP_INFO(parent_node->get_logger(), "Converting GPS waypoints to Map Frame..");
+  rclcpp::Duration transform_tolerance(0, 500);
+
+  std::vector<geometry_msgs::msg::PoseStamped> poses_in_map_frame_vector;
+  for (auto && curr_gps_pose : gps_poses) {
+    auto request = std::make_shared<robot_localization::srv::FromLL::Request>();
+    auto response = std::make_shared<robot_localization::srv::FromLL::Response>();
+    request->ll_point.latitude = curr_gps_pose.position.latitude;
+    request->ll_point.longitude = curr_gps_pose.position.longitude;
+    request->ll_point.altitude = curr_gps_pose.position.altitude;
+
+    fromll_client->wait_for_service((std::chrono::seconds(1)));
+    auto is_conversion_succeeded = fromll_client->invoke(
+      request,
+      response);
+    if (!is_conversion_succeeded) {
+      RCLCPP_ERROR(
+        parent_node->get_logger(),
+        "fromLL service of robot_localization could not convert %i th GPS waypoint to"
+        "Map frame, going to skip this point!"
+        "Make sure you have run navsat_transform_node of robot_localization");
+      continue;
+    } else {
+      // this poses are assumed to be on global frame (map)
+      geometry_msgs::msg::PoseStamped curr_pose_in_map_frame;
+      curr_pose_in_map_frame.header.frame_id = "map";
+      curr_pose_in_map_frame.header.stamp = parent_node->now();
+      curr_pose_in_map_frame.pose.position.x = response->map_point.x;
+      curr_pose_in_map_frame.pose.position.y = response->map_point.y;
+      curr_pose_in_map_frame.pose.position.z = response->map_point.z;
+
+      geometry_msgs::msg::Quaternion utm_orientation = curr_gps_pose.orientation;
+      geometry_msgs::msg::PoseStamped temporary_pose;
+      temporary_pose.pose.orientation = utm_orientation;
+      temporary_pose.header.frame_id = "utm";
+      nav_2d_utils::transformPose(
```

Comment:
Ah that's why you grabbed nav 2d utils, please use TF directly or if you want to create a new utility for it, that's OK too. but I think the simple `tf->transform(in_pose, out_pose, frame, transform_timeout);` in a `try` is fine. Alot of those other exceptions in that file are really not ever triggered worth the complexity.

Reply:
Did you test that varying the orientation change the robot behavior? Just want validation that the orientations are being actually used in the waypoints (e.g. when you follow GPS waypoints with different ending orientations, do you see the robot using different final orientations?)

Reply:
Yes I tested, also see here https://github.com/ros-planning/navigation2/pull/2111#discussion_r543220857